File: papers/journalEM/code/asiapns.py
#!/usr/bin/python3


import subprocess
import datetime
import os
import sys


from datetime import datetime
from pathlib import Path

#### Parameter experiments

print(sys.argv)


cause,alg="asia","CCALP"
cause = str(sys.argv[1])
alg = str(sys.argv[2])
heapGB = 64


####

def gen_exec(cmd, check_return: bool = False):
    popen = subprocess.Popen(cmd, stdout=subprocess.PIPE, universal_newlines=True)
    for stdout_line in iter(popen.stdout.readline, ""):
        yield stdout_line
    popen.stdout.close()
    return_code = popen.wait()
    if return_code and check_return:
        raise subprocess.CalledProcessError(return_code, cmd)


def exec_bash(cmd: str, check_return: bool = False):
    return [s for s in gen_exec(cmd.split(), check_return)]
    
def exec_bash_print(cmd: str, check_return: bool = False):
    for path in gen_exec(cmd.split(), check_return):
        print(path, end="")
        
def strtime():
    return datetime.now().strftime("%y%m%d_%H%M%S")

prj_path = Path("/Users/rcabanas/GoogleDrive/IDSIA/causality/dev/credici/")
prj_path = Path(str(Path("../../../").resolve())+"/")
exp_folder = Path(prj_path, "papers/journalEM/")
code_folder = Path(exp_folder, "code")
res_folder = Path(exp_folder, "output")
model_folder = Path(exp_folder, "models")
data_folder = Path(exp_folder, "data")


jar_file = Path(prj_path, "target/credici-0.1.5-dev-SNAPSHOT-jar-with-dependencies.jar")
#java = "/Library/Java/JavaVirtualMachines/openjdk-12.0.1.jdk/Contents/Home/bin/java"
java = "java"


print(prj_path)
print(exp_folder)
print(res_folder)
print(model_folder)
print(jar_file)


def runjava(javafile, args_str, heap_gbytes=None):
    print("run java")
    cmd = f"{java} "
    if heap_gbytes is not None: cmd +=f"-Xmx{heap_gbytes}g "
    cmd += f"-cp {jar_file} {javafile} {args_str}"
    print(cmd)
    exec_bash_print(cmd)

## Get models

# -w
# -x 100
# -m 500 -sc KL -th 0.00001 -a EMCC -rw --seed 0 --output ./papers/journalEM/output/synthetic/sample_files/ ./papers/journalEM/models/synthetic/s1/random_mc2_n6_mid3_d1000_05_mr098_r10_17.uai

# -x 5 -m 10 --debug -a CCVE --cause asia -rw -o /Users/rcabanas/GoogleDrive/IDSIA/causality/dev/credici/papers/journalEM/output/literature --seed 0
# /Users/rcabanas/GoogleDrive/IDSIA/causality/dev/credici/papers/journalEM/models/literature/


def asiapns(method, cause,
             rewrite = True,
             executions = 300,
             max_iter = 500,
             output = ".", seed = 0, debug=False):


    args = ""
    if rewrite: args += f"-rw "
    if debug: args += f"--debug "

    args += f"-x {executions} "
    args += f"-m {max_iter} "
    args += f"-a {method} "
    args += f"--seed {seed} "
    args += f"--output {output} "
    if cause is not None: args += f"--cause {cause} "
    args += f"{Path(model_folder, './literature/')}"

    javafile = Path(code_folder, "AsiaPNS.java")
    print(javafile)
    runjava(javafile, args_str=args, heap_gbytes=heapGB)

####

asiapns(alg, cause, output=Path(res_folder, "./literature"), debug=True)

File: papers/journalEM/code/AsiaPNS.java
package code;

import ch.idsia.credici.IO;
import ch.idsia.credici.inference.CausalEMVE;
import ch.idsia.credici.inference.CausalInference;
import ch.idsia.credici.inference.CredalCausalApproxLP;
import ch.idsia.credici.inference.CredalCausalVE;
import ch.idsia.credici.learning.FrequentistCausalEM;
import ch.idsia.credici.model.StructuralCausalModel;
import ch.idsia.credici.model.builder.CausalBuilder;
import ch.idsia.credici.utility.CollectionTools;
import ch.idsia.credici.utility.Probability;
import ch.idsia.credici.utility.experiments.ResultsManager;
import ch.idsia.credici.utility.experiments.Terminal;
import ch.idsia.credici.utility.experiments.Watch;
import ch.idsia.crema.factor.GenericFactor;
import ch.idsia.crema.factor.convert.VertexToInterval;
import ch.idsia.crema.factor.credal.linear.IntervalFactor;
import ch.idsia.crema.factor.credal.vertex.VertexFactor;
import ch.idsia.crema.model.graphical.specialized.BayesianNetwork;
import ch.idsia.crema.utility.RandomUtil;
import gnu.trove.map.TIntIntMap;
import jdk.jshell.spi.ExecutionControl;
import picocli.CommandLine;

import java.io.File;
import java.io.IOException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.DoubleStream;

public class AsiaPNS extends Terminal {

    /*

    -x 5 -m 50 -a EMCC /Users/rcabanas/GoogleDrive/IDSIA/causality/dev/credici/papers/journalEM/models/literature/

     */

    @CommandLine.Parameters(description = "Folder with the model asia.uai")
    private String modelPath;


    @CommandLine.Option(names = {"-m", "--maxiter"}, description = "Maximum EM internal iterations. Default to 500")
    private int maxIter = 500;


    @CommandLine.Option(names={"-rw", "--rewrite"}, description = "If activated, results are rewritten. Otherwise, process is stopped if there are existing results.")
    boolean rewrite = false;

    @CommandLine.Option(names={"-o", "--output"}, description = "Output folder for the results. Default working dir.")
    String output = ".";

    @CommandLine.Option(names = {"-x", "--executions"}, description = "Number independent EM runs. Only for EM-based methods. Default to 40")
    private int executions = 40;


    @CommandLine.Option(names = {"-a", "--algorithm"}, description = "Learning and inference algorithm: ${COMPLETION-CANDIDATES}")
    private AsiaPNS.algorithms alg = AsiaPNS.algorithms.CCVE;


    public enum algorithms {
        CCVE,   // Credal causal VE
        CCALP,  // Credal causal approx LP
        EMCC,   // EM

    }


    @CommandLine.Option(names = {"-c", "--cause"}, description = "Cause variable. Default to smoke. Possible values: smoke, bronc, dysp, either, lung, tub, asia, xray")
    private String cause = "smoke";
    private String effect = "dysp";

    // Global Variables
    String[] varnames = {"smoke","bronc","dysp","either","lung","tub","asia","xray"};
    TIntIntMap[] data = null;
    StructuralCausalModel model = null;
    int yes = 0;
    int no = 1;

    ResultsManager res = null;

    // Outputs
    List<Double> pnsPoints = new ArrayList<>();
    List<Double> llk = new ArrayList<>();
    double maxllk = Double.NaN;
    double lb = Double.NaN;
    double ub = Double.NaN;
    double timelearn = Double.NaN;
    double  timeinference = Double.NaN;

    private void processResults() throws IOException {
        res.add("pns", CollectionTools.toDoubleArray(pnsPoints));
        res.add("llk", CollectionTools.toDoubleArray(llk));
        res.add("pns_u", ub);
        res.add("pns_l", lb);
        res.add("time_learn", timelearn);
        res.add("time_inf", timeinference);
        res.add("cause", cause);
        res.add("effect", effect);
        res.add("method", alg.toString());
        res.add("max_llk", maxllk);

        res.logSummary();

        Path fullpath = Path.of(output, getLabel()+".csv");

        if(rewrite || !fullpath.toFile().exists())
            res.save(fullpath.toAbsolutePath().toString());

    }


    @Override
    protected void entryPoint() throws Exception {

        init();
        loadModel();
        runInference();
        processResults();   //todo: set result path

    }

    protected void init(){
        res = new ResultsManager()
                .setIncludeLabel(false).setLogger(logger);
    }

    protected String getLabel(){
        return getLabel(this.alg);
    }

    private String getLabel(algorithms alg) {
        //mIter500_wtrue_sparents3_x20_0
        String str = "asia";

        str+="_"+alg;

        if(alg == algorithms.EMCC) {
            str += "_mIter" + this.maxIter;
            str += "_x" + this.executions;
        }
        str += "_"+cause;
        str += "_"+effect;



        if(alg != algorithms.CCVE) {
            str += "_" + this.seed;
        }

        return str;
    }
    private void runInference() throws InterruptedException, ExecutionControl.NotImplementedException {

        int y = List.of(varnames).indexOf("dysp");
        int x = List.of(varnames).indexOf(cause);

        logger.debug("X="+x+" ("+cause+"), Y="+y+" ("+effect+")");

        maxllk = Probability.maxLogLikelihood(model,data);

        logger.debug("Max llk: "+maxllk);

        String method = alg.toString();

        CausalInference inf = null;

        logger.info("Starting learning");

        Watch.start();

        if(method=="CCVE")
            inf = new CredalCausalVE(model, data);
        else if(method=="CCALP")
            inf = new CredalCausalApproxLP(model,data);
        else if(method=="EMCC")
            inf = new CausalEMVE(model, data, executions, maxIter);
        else
            throw new IllegalArgumentException("Unknown inference method");

        Watch.stopAndLog(this.logger, "Learning finished in ");
        timelearn = Watch.getTime();

        Watch.start();

        if(method != "EMCC") {
            GenericFactor res = inf.probNecessityAndSufficiency(x, y, yes, no);
            if (!(res instanceof IntervalFactor)) {
                res = new VertexToInterval().apply((VertexFactor) res, y);
            }
            lb = ((IntervalFactor) res).getDataLower()[0][0];
            ub = ((IntervalFactor) res).getDataUpper()[0][0];
            pnsPoints.add(lb);
            pnsPoints.add(ub);
        }else{
            double pns[] = (((CausalEMVE) inf).getIndividualPNS(x, y, yes, no));
            pnsPoints = DoubleStream.of(pns).boxed().collect(Collectors.toList());
            ub = Arrays.stream(pns).max().getAsDouble();
            lb = Arrays.stream(pns).min().getAsDouble();
            llk = ((CausalEMVE) inf).getInputModels().stream().map(m -> m.logLikelihood(data)).collect(Collectors.toList());

            logger.debug("PNS: "+pnsPoints);
            logger.debug("llk: "+llk);


        }

        Watch.stopAndLog(this.logger, "Inference finished  in ");
        timeinference = Watch.getTime();


        logger.info("PNS(" + cause + ","+effect+") = [" + lb + "," + ub + "]\t ("+method+")");



    }

    private void loadModel() throws IOException {

        String path = Path.of(modelPath, "asia.uai").toString();

        logger.info("Loading model from"+path);
        // Load the data and the model
        BayesianNetwork bnet = (BayesianNetwork) IO.readUAI(path);

        RandomUtil.setRandomSeed(0);
        data = bnet.samples(1000, bnet.getVariables());

        RandomUtil.setRandomSeed(seed);
        model = CausalBuilder.of(bnet.getNetwork(), 2).build();

        logger.info("Built markovian model");


    }


    public static void main(String[] args) {
        argStr = String.join(";", args);
        CommandLine.run(new AsiaPNS(), args);
        if(errMsg!="")
            System.exit(-1);
        System.exit(0);
    }


}
